nomos-services/chain-service/src/lib.rs

```diff
@@ -1644,3 +1656,197 @@ where
         .map_err(|(error, _)| Box::new(error) as DynError)?;
     receiver.await.map_err(|error| Box::new(error) as DynError)
 }
+
+/// Reconstruct a Block from a Proposal by looking up transactions from mempool
+async fn reconstruct_block_from_proposal<Payload, Item, Key>(
+    proposal: Proposal,
+    mempool: OutboundRelay<MempoolMsg<HeaderId, Payload, Item, Key>>,
+) -> Result<Block<Item>, Error>
+where
+    Key: Send + Eq + Hash + From<ZkHash> + Debug,
+    Payload: Send,
+    Item: Clone + Transaction<Hash = Key>,
+{
+    let tx_map: HashMap<Key, Item> = get_mempool_contents(mempool)
+        .await
+        .map_err(|e| Error::BlockReconstruction(format!("Failed to get mempool contents: {e}")))?
+        .map(|tx| (tx.hash(), tx))
+        .collect();
+
+    let mut reconstructed_transactions = Vec::new();
+    let mut missing_transactions = Vec::new();
+
+    for tx_ref in &proposal.references().mempool_transactions {
+        let tx_hash = Key::from(*tx_ref);
+
+        match tx_map.get(&tx_hash) {
+            Some(transaction) => {
+                reconstructed_transactions.push(transaction.clone());
+            }
+            None => {
+                missing_transactions.push(tx_hash);
+            }
+        }
+    }
+
+    if !missing_transactions.is_empty() {
+        return Err(Error::BlockReconstruction(format!(
+            "Failed to reconstruct block: {missing_transactions:?} transactions not found in mempool",
+        )));
+    }
+
+    let header = proposal.header().clone();
+
+    Ok(Block::new(header, reconstructed_transactions))
```

Comment:
we need a lot more validation here, but this was not the scope for this PR :), I'd say just put a comment here that we miss validation